Feature/docsite #15 (#16)
zeppelinos/zeppelin-solidity-hot-loader: docs/modules/ROOT/pages/getting-started.adoc

= Getting Started

== Introduction

The easiest way to get started using Solidity Hot Loader is using a xref:starter-kits::index.adoc[Starter Kit]. In this guide we walk through using Solidity Hot Loader with the xref:starter-kits::tutorial.adoc[Tutorial Starter Kit]. 

This Kit includes a Counter dapp, which displays the value of a Counter that can be increased and decreased: we will make changes to the code of the smart contract to show the power of the Solidity Hot Loader.

== Setup

> First we need a xref:learn::deploy-and-interact.adoc#local-blockchain[local blockchain] to develop on. We will use `ganache-cli`, so go ahead and install (if you haven't already) and then run a local blockchain using `ganache-cli`

Reviewer: ```suggestion
First we need a xref:learn::deploying-and-interacting.adoc#local-blockchain[local blockchain] to develop on. We will use `ganache-cli`, so go ahead and install (if you haven't already) and then run a local blockchain using `ganache-cli`
```

[source,console]
----
$ npm install -g ganache-cli
$ ganache-cli --deterministic
----

Next we need a project directory to develop our project in. In a separate terminal create the project directory:

[source,console]
----
$ mkdir myproject && cd myproject
----

> Unpack the Tutorial Starter Kit using the xref:sdk::index.adoc[*OpenZeppelin CLI*]. This could take a little while, so maybe make a cup of tea.

Reviewer: ```suggestion
Unpack the Tutorial Starter Kit using the xref:cli::index.adoc[*OpenZeppelin CLI*]. This could take a little while, so maybe make a cup of tea.
```

? Initial project version 1.0.2
Project initialized. Write a new contract in the contracts folder and run 'openzeppelin create' to deploy it.
----

The Tutorial Starter Kit already comes with a `Counter` contract: compile it using the OpenZeppelin CLI:

[source,console]
----
$ npx openzeppelin compile
✓ Compiled contracts with solc 0.5.13 (commit.5b0b510c)
----

We will now deploy `Counter` to our local ganache blockchain. Run `npx openzeppelin create` and use the interactive commands to deploy the `Counter` contract to the development network, calling `initialize` with a numeric value (here we use '23'):

[source,console]
----
$ npx openzeppelin create
Nothing to compile, all contracts are up to date.
? Pick a contract to instantiate Counter
? Pick a network development
✓ Added contract Counter
✓ Contract Counter deployed
All contracts have been deployed
? Call a function to initialize the instance after creating it? Yes
? Select which function * initialize(num: uint256)
? num (uint256): 23
✓ Setting everything up to create contract instances
✓ Instance created at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601
0xCfEB869F69431e42cdB54A4F4f105C19C080A601
----

> TIP: If you're unfamiliar with using the OpenZeppelin CLI to deploy smart contracts, head to our xref:learn::deploy-and-interact.adoc#getting-started-with-the-cli[Getting Started] guide.

Reviewer: ```suggestion
TIP: If you're unfamiliar with using the OpenZeppelin CLI to deploy smart contracts, head to our xref:learn::deploying-and-interacting.adoc#getting-started-with-the-cli[Getting Started] guide.
```

= Getting Started

== Introduction

> The easiest way to get started using Solidity Hot Loader is using a https://docs.openzeppelin.com/starter-kits/2.3/[Starter Kit]. In this guide we walk through using Solidity Hot Loader with the https://docs.openzeppelin.com/starter-kits/2.3/tutorial[Tutorial Starter Kit]. The Starter Kit includes a Counter dapp, which displays the value of a Counter that can be increased/decreased. We will update the code in the smart contract to show the power of the Solidity Hot Loader.

Reviewer: ```suggestion
The easiest way to get started using Solidity Hot Loader is using a xref:starter-kits::index.adoc[Starter Kit]. In this guide we walk through using Solidity Hot Loader with the xref:starter-kits::tutorial.adoc[Tutorial Starter Kit]. 

This Kit includes a Counter dapp, which displays the value of a Counter that can be increased and decreased: we will make changes to the code of the smart contract to show the power of the Solidity Hot Loader.
```

== Setup

> First we need a local blockchain to develop on. We use `ganache-cli`, so go ahead and install (if you haven't already) and then run a local blockchain using `ganache-cli`

Reviewer: ```suggestion
First we need a xref:learn::deploy-and-interact.adoc#local-blockchain[local blockchain] to develop on. We will use `ganache-cli`, so go ahead and install (if you haven't already) and then run a local blockchain using `ganache-cli`
```

[source,console]
----
$ npm install -g ganache-cli
$ ganache-cli --deterministic
----

> Next we need a project directory to develop our project in. In a separate terminal create the project directory.

Reviewer: ```suggestion
Next we need a project directory to develop our project in. In a separate terminal create the project directory:
```

[source,console]
----
$ mkdir myproject && cd myproject
----

> Unpack the Tutorial Starter Kit. (This could take a little while so maybe make a cup of tea)

Reviewer: ```suggestion
Unpack the Tutorial Starter Kit using the xref:sdk::index.adoc[*OpenZeppelin CLI*]. This could take a little while, so maybe make a cup of tea.
```

[source,console]
----
$ npx @openzeppelin/cli unpack tutorial
----

> Initialize the project with OpenZeppelin.

Reviewer: I think this isn't needed yet, I'd postpone this until we use `oz create` to not bog down readers with one-off commands here.

Author: I thought we needed it before compiling, so have moved to before compile.

[source,console]
----
$ npx openzeppelin init
? Welcome to the OpenZeppelin SDK! Choose a name for your project myproject
? Initial project version 1.0.2
Project initialized. Write a new contract in the contracts folder and run 'openzeppelin create' to deploy it.
----

> Enable the Solidity Hot Loader in `client/config/webpack.js` by setting `disabled: false` in your favorite editor. We can disable this when we need.

Reviewer: Does this file already exist when downloading the kit, or do users need to create it?

Author: The file already exists, so we could cut down to just show the `disabled: false`

> Enable the Solidity Hot Loader in `client/config/webpack.js` by setting `disabled: false` in your favorite editor. We can disable this when we need.

Reviewer: ```suggestion
The Solidity Hot Loader already comes installed with the Tutorial Starter Kit: all we need to do is enable it. Do this by setting `disabled: false` in `client/config/webpack.js`:
```

[source,js]
----

> const solidityLoaderOptions = {

Reviewer: ```suggestion
// client/config/webpack.js

const solidityLoaderOptions = {
```

We've been following a convention of adding the name of the file at the beginning of each snippet, which I find helps a lot understand what the contents refer to

Author: Added

  network: 'development',
  // you can stop loader from automatic compile/push/upgrade
  // action by setting disabled flag to true, but it will still
  // serve .json files from file system
  disabled: false,
};

module.exports = {
  solidityLoader: {
    test: /\.sol$/,
    use: [
      { loader: 'json-loader' },
      {
        loader: '@openzeppelin/solidity-loader',
        options: solidityLoaderOptions,
      },
    ],
  },
  solidityLoaderOptions,
};
----

> == Counter contract

Reviewer: ```suggestion
== Compiling and Deploying
```

> Compile the Counter contract with OpenZeppelin CLI.

Reviewer: ```suggestion
The Tutorial Starter Kit already comes with a `Counter` contract: compile it using the OpenZeppelin CLI:
```

[source,console]
----
$ npx oz compile
✓ Compiled contracts with solc 0.5.13 (commit.5b0b510c)
----

> Deploy the Counter contract to our local blockchain.

Reviewer: ```suggestion
```

> Run `npx oz create` and use the interactive commands to deploy the Counter contract to the development network and initialize to a value e.g. `23`.

Reviewer: ```suggestion
We will now deploy `Counter` to our local ganache blockchain. Run `npx oz create` and use the interactive commands to deploy the `Counter` contract to the development network, calling `initialize` with a numeric value (here we use '23'):
```

Author: Updated

[source,console]
----
$ npx oz create
Nothing to compile, all contracts are up to date.
? Pick a contract to instantiate Counter
? Pick a network development
✓ Added contract Counter
✓ Contract Counter deployed
All contracts have been deployed
? Call a function to initialize the instance after creating it? Yes
? Select which function * initialize(num: uint256)
? num (uint256): 23
✓ Setting everything up to create contract instances
✓ Instance created at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601
0xCfEB869F69431e42cdB54A4F4f105C19C080A601
----

> 

Reviewer: ```suggestion
TIP: If you're unfamiliar with using the OpenZeppelin CLI to deploy smart contracts, head to our xref:learn::deploy-and-interact.adoc#getting-started-with-the-cli[Getting Started] guide.

```

> == App

Reviewer: ```suggestion
== Using our dapp
```

> We will now start the app.

Reviewer: ```suggestion

We are now ready to start the dapp!
```

[source,console]
----
$ cd client
$ npm run start
----

> The app will open in a browser on http://localhost:3000/

Reviewer: Consider merging these sentences into a single paragraph so that the text isn't too broken up

Author: Updated (slightly)

Click on Counter to go to http://localhost:3000/counter

In MetaMask change your network to Localhost:8545 to connect to `ganache-cli`.

== Fund MetaMask account
If you need to fund your MetaMask account (e.g. it has a balance of 0 test Ether).

From your project directory in another terminal, call `npx oz transfer` to transfer 0.5 (test) ether to your account in MetaMask.

[source,console]
----
$ npx oz transfer
? Pick a network development
? Choose the account to send transactions from (0) 0xF983502f33E2c9E62d0465557446B86C595C4919
? Enter the receiver account 0x77737a65C296012C67F8c7f656d1Df81827c9541
? Enter an amount to transfer 0.5 ether
✓ Funds sent. Transaction hash: 0x69a0ecc17b325a0e4cd414ca9d197a59e88a466c8d3b4d7a0a518ca384526af4
----

== Use the app

> In the app, press *Increase Counter by 1* button. Confirm the transaction in MetaMask and once the transaction is confirmed the app will have the updated counter value.

Reviewer: I'd rephrase this and the following section to explain a bit more what's going on, instead of just giving instructions. For example after running the app we could go 'and what if we now wanted to change the contract? we'd have to do x and y and z, rerun our app, etc', so that the added value of the loader is better conveyed.

Author: Updated

== Update the Counter contract
In an editor, modify the Counter contract to change the calculation in the `increaseCounter` function to add 3.

[source,solidity]
----
  function increaseCounter(uint256 amount) public {
    count = count + 3;
  }
----

The Solidity Hot Loader will recompile the contract and upgrade it on the development network and update the artifacts. We can see in the terminal when it is compiling.

In the app, notice that the state doesn't change. Then press *Increase Counter by 1* button. Confirm the transaction in MetaMask and once the transaction is confirmed the app will have the updated counter value. 
The value should increase by 3.

== Review

> Let's take a moment and review the Solidity Hot Loader development cycle: 

Reviewer: I'd avoid repeating the content as-is from the overview (though the underlying idea is great)

Author: Removed review section
